File: linked-list-medium/92.反转链表-ii.py
#
# @lc app=leetcode.cn id=92 lang=python3
#
# [92] 反转链表 II
#

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @lc code=start
class Solution:
    # 44/44 cases passed (56 ms)
    # Your runtime beats 5.76 % of python3 submissions
    # Your memory usage beats 20.37 % of python3 submissions (15.1 MB)
    def reverseN(self, head: ListNode, N: int) -> ListNode:
        if N == 1:
            self.join = head.next
            return head
        last = self.reverseN(head.next, N - 1)
        head.next.next = head
        head.next = self.join
        return last

    def reverseBetween(self, head: ListNode, left: int, right: int) -> ListNode:
        if left == 1:
            return self.reverseN(head, right)

        head.next = self.reverseBetween(head.next, left - 1, right - 1)
        return head


# @lc code=end


class Solution_1:
    # 执行用时： 56 ms , 在所有 Python3 提交中击败了 5.72% 的用户
    # 内存消耗： 14.9 MB , 在所有 Python3 提交中击败了 67.80% 的用户
    def reverseBetween(self, head: ListNode, left: int, right: int) -> ListNode:
        dummy = ListNode()
        dummy.next = head
        pre = dummy
        # Let pre is (left - 1) node
        for _ in range(left - 1):
            pre = pre.next
        # Let cur is (left) node
        cur = pre.next
        for _ in range(right - left):
            tmp = pre.next
            pre.next = cur.next
            cur.next = cur.next.next
            pre.next.next = tmp
        # Note: not return head:
        # eg.[3, 5], left=1, right=2, return head will cause reversed list to be [3]
        return dummy.next

File: linked-list-medium/92.反转链表-ii.cpp
/*
 * @lc app=leetcode.cn id=92 lang=cpp
 *
 * [92] 反转链表 II
 */

/**
 * Definition for singly-linked list.
 */

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

// @lc code=start
class Solution {
    // 44/44 cases passed (4 ms)
    // Your runtime beats 51.75 % of cpp submissions
    // Your memory usage beats 20.43 % of cpp submissions (7.3 MB)
    ListNode* successor = nullptr;
    ListNode* reverseN(ListNode* head, int n) {
        if (n == 1) {
            successor = head->next;
            return head;
        }
        ListNode* last = reverseN(head->next, n - 1);
        head->next->next = head;
        head->next = successor;
        return last;
    }

   public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (m == 1) {
            return reverseN(head, n);
        }
        head->next = reverseBetween(head->next, m - 1, n - 1);
        return head;
    }
};
// @lc code=end

class Solution_1 {
   public:
    // 44/44 cases passed (0 ms)
    // Your runtime beats 100 % of cpp submissions
    // Your memory usage beats 13.35 % of cpp submissions (7.3 MB)
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode *dummy = new ListNode(), *pre = dummy, *cur;
        dummy->next = head;
        for (int i = 0; i < m - 1; i++) {
            pre = pre->next;
        }
        cur = pre->next;
        for (int i = 0; i < n - m; i++) {
            ListNode* temp = pre->next;
            pre->next = cur->next;
            cur->next = cur->next->next;
            pre->next->next = temp;
        }
        return dummy->next;
    }
};